### PR TITLE
HIVE-22369 Handle HiveTableFunctionScan at return path

### PR DESCRIPTION
**Pull request by miklosgergely**
The optimizedOptiqPlan at CalcitePlanner.getOptimizedHiveOPDag is ultimately generated by CalcitePlanner.internalGenSelectLogicalPlan, which may either provide a HiveProject or a HiveTableFunctionScan. When HiveCalciteUtil.getTopLevelSelect is invoked on this it is looking for a HiveProject node in the tree, which if won't find in case of a HiveTableFunctionScan was returned. This is why TestNewGetSplitsFormat is failing with return path.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/HiveOpConverter.java [hunk 1]
```diff
@@ -186,12 +193,67 @@ OpAttr dispatch(RelNode rn) throws SemanticException {
...
+    op.getConf().setRowLimit(1);
+    qb.addAlias(SemanticAnalyzer.DUMMY_TABLE);
+    qb.setTabAlias(SemanticAnalyzer.DUMMY_TABLE, SemanticAnalyzer.DUMMY_TABLE);
+
+    Operator<?> output = OperatorFactory.getAndMakeChild(new SelectDesc(exprCols, fieldNames, false),
+        new RowSchema(rowResolver.getRowSchema()), op);
+    output.setColumnExprMap(colExprMap);
+    semanticAnalyzer.putOpInsertMap(output, rowResolver);
+
+    Operator<?> funcOp = semanticAnalyzer.genUDTFPlan(genericUDTF, null, fieldNames, qb, output, false);
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/HiveOpConverter.java]: Should we move this method to `HiveOpConverter` and clean any context information that is not needed? Similar to `genReduceSink`. Idea is that when we enable this path, we can easily get rid of most dependencies from `SemanticAnalyzer`.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/HiveOpConverter.java [hunk 2]
```diff
@@ -186,12 +193,67 @@ OpAttr dispatch(RelNode rn) throws SemanticException {
...
+      colExprMap.put(exprNames.get(pos), exprCol);
+      exprCols.add(exprCol);
+
+      ColumnInfo columnInfo = new ColumnInfo(fieldNames.get(pos), exprCol.getWritableObjectInspector(), null, false);
+      rowResolver.put(columnInfo.getTabAlias(), columnInfo.getAlias(), columnInfo);
+    }
+
+    QB qb = new QB(semanticAnalyzer.getQB().getId(), nextAlias(), true);
+    qb.getMetaData().setSrcForAlias(SemanticAnalyzer.DUMMY_TABLE, semanticAnalyzer.getDummyTable());
+    TableScanOperator op = (TableScanOperator) semanticAnalyzer.genTablePlan(SemanticAnalyzer.DUMMY_TABLE, qb);
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/HiveOpConverter.java]: We could create the operator in this class (we could add a method `createDummyTableScan`). It should probably be a subset of `visit(HiveTableScan scanRel)`? It could also be useful for other translations. That way we can also remove the dependency to QB, which we want eventually to get rid of in this path.

Comment (miklosgergely) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/HiveOpConverter.java]: QB will be removed completely? I'll check out how we could integrate this to `visit(HiveTableScan scanRel)`.

**Comment by miklosgergely** [comment]
Added unit tests for the modifications.

**Comment by miklosgergely** [comment]
@jcamachor, I think I understood what you asked for, please check out the new patch, and let me know if this is how you wanted it.

File: itests/hive-unit/src/test/java/org/apache/hive/jdbc/BaseJdbcWithMiniLlap.java [hunk 3]
```diff
@@ -530,12 +498,10 @@ protected int processQuery(String currentDatabase, String query, int numSplits,
     InputSplit[] splits = inputFormat.getSplits(job, numSplits);
 
     // Fetch rows from splits
-    boolean first = true;
```
Comment (jcamachor) [inline itests/hive-unit/src/test/java/org/apache/hive/jdbc/BaseJdbcWithMiniLlap.java]: These changes do not seem related to this patch? If they are not, can we submit it in a different one?
